File: decorator/Decorator.py
from typing import Protocol

class IWeapon(Protocol):
    def describe():
        pass
    def getAttack()->int:
        pass

class Sword(IWeapon):
    def describe(self):
        print("This is a sword")
    def getAttack(self)->int:
        return 10

class IWeaponDecorator(IWeapon):
    pass

class SuperSwordDecorator(IWeaponDecorator):
    def __init__(self, weapon:IWeapon):
        self.weapon = weapon
    def describe(self):
        print("This is a super version")
        self.weapon.describe()
    def getAttack(self)->int:
        return self.weapon.getAttack() + 20
    
def testDecorator():
    sword = Sword()
    sword.describe()
    print("Attack:", sword.getAttack())
    superSword = SuperSwordDecorator(sword)
    superSword.describe()
    print("Attack:", superSword.getAttack())

if __name__ == "__main__":
    testDecorator()

File: decorator/Decorator.ts
interface IWeapon {
    describe(): void;
    getAttack(): number;
}

class Sword implements IWeapon {
    public describe(): void {
        console.log("This is a sword");
    }

    public getAttack(): number {
        return 20;
    }
}

interface IWeaponDecorator extends IWeapon { };

class SuperSwordDecorator implements IWeaponDecorator {
    private weapon: IWeapon;

    constructor(weapon: IWeapon) {
        this.weapon = weapon;
    }

    public describe(): void {
        console.log("This is a super sword");
        this.weapon.describe();
    }

    public getAttack(): number {
        return this.weapon.getAttack() + 10;
    }
}

function testDecorator() {
    const sword = new Sword();
    sword.describe();
    console.log("Attack: ", sword.getAttack());

    const superSword = new SuperSwordDecorator(sword);
    superSword.describe();
    console.log("Attack: ", superSword.getAttack());
}

testDecorator();